api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -495,25 +534,38 @@ def __iter__(self) -> Iterator[ObsAndXDatum]:
                 "(see https://github.com/pytorch/pytorch/issues/20248)"
             )
 
-        with _open_experiment(self.exp_uri, self.aws_region, soma_buffer_bytes=self.soma_buffer_bytes) as exp:
-            X: soma.SparseNDArray = exp.ms[self.measurement_name].X[self.layer_name]
+        # chunk the obs joinids into batches of size soma_chunk_size
+        obs_joinids_chunked = self._chunk_ids(self._obs_joinids, self.soma_chunk_size)
 
-            # TODO: Get all IDs, close experiment. Iterate through obs IDs using custom stride, not based upon soma_buffer_bytes
-            obs_tables_iter = exp.obs.read(
-                coords=(self._partition_obs_joinids(self._obs_joinids),),
-                batch_size=somacore.BatchSize(),
-                column_names=self.obs_column_names,
-            )
+        # globally shuffle the chunks, if requested
+        if self.shuffle:
+            self._global_shuffle_obs_joinids(obs_joinids_chunked)
+
+        # subset to a single partition, as needed for distributed training and multi-processing datat loading
+        worker_info = torch.utils.data.get_worker_info()
+        partition, partitions = self._compute_partitions(
+            loader_partition=worker_info.id if worker_info else 0,
+            loader_partitions=worker_info.num_workers if worker_info else 1,
+            dist_partition=dist.get_rank() if dist.is_initialized() else 0,
+            num_dist_partitions=dist.get_world_size() if dist.is_initialized() else 1,
+        )
+        obs_joinids_chunked_partition: npt.NDArray[np.int64] = self._subset_ids_to_partition(
+            obs_joinids_chunked, partition, partitions
+        )
 
+        with _open_experiment(self.exp_uri, self.aws_region) as exp:
             obs_and_x_iter = _ObsAndXIterator(
-                X,
-                obs_tables_iter,
+                obs=exp.obs,
+                X=exp.ms[self.measurement_name].X[self.layer_name],
+                obs_column_names=self.obs_column_names,
+                obs_joinids_chunked=obs_joinids_chunked_partition,
```

Comment:
We now pass in the obs ids rather than a obs_tables_iter (a SOMA Reader). The id are chunked, optionally shuffled, and optionally partitioned.